macosx/QuickLookExtension/PreviewProvider.mm

```diff
@@ -0,0 +1,290 @@
+        NSImage* renderedIcon = [[NSImage alloc] initWithSize:iconFrame.size];
+        [renderedIcon lockFocus];
+        [icon drawInRect:iconFrame];
+        [renderedIcon unlockFocus];
+
+        NSData* iconData = renderedIcon.TIFFRepresentation;
+
+        QLPreviewReplyAttachment* imageAttachment = [[QLPreviewReplyAttachment alloc] initWithData:iconData contentType:UTTypePNG];
+
+        allImgProps[iconFileName] = imageAttachment;
+    }
+
+    return [@"cid:" stringByAppendingString:iconFileName];
+}
+
+@implementation PreviewProvider
+
+- (void)providePreviewForFileRequest:(QLFilePreviewRequest*)request
+                   completionHandler:(void (^)(QLPreviewReply* _Nullable reply, NSError* _Nullable error))handler
+{
+    QLPreviewReply* reply = [[QLPreviewReply alloc]
+        initWithDataOfContentType:UTTypeHTML
+                      contentSize:CGSizeMake(1200, 800)
+                dataCreationBlock:^NSData* _Nullable(QLPreviewReply* _Nonnull replyToUpdate, NSError* __autoreleasing _Nullable* _Nullable error) {
+                    NSString* previewHTML = [self generateHTMLPreviewFor:request.fileURL andReply:replyToUpdate];
+
+                    return [previewHTML dataUsingEncoding:NSUTF8StringEncoding];
+                }];
+
+    handler(reply, nil);
+}
+
+- (NSString*)generateHTMLPreviewFor:(NSURL*)url andReply:(QLPreviewReply* _Nonnull)replyToUpdate
+{
+    // Try to parse the torrent file
+    auto metainfo = tr_torrent_metainfo{};
+    if (!metainfo.parse_torrent_file(url.path.UTF8String))
+    {
+        return nil;
+    }
+
+    NSBundle* bundle = [NSBundle mainBundle];
+
+    NSURL* styleURL = [bundle URLForResource:@"style" withExtension:@"css"];
+    NSString* styleContents = [NSString stringWithContentsOfURL:styleURL encoding:NSUTF8StringEncoding error:NULL];
+
+    NSMutableString* htmlString = [NSMutableString string];
+    [htmlString appendFormat:@"<html><style type=\"text/css\">%@</style><body>", styleContents];
+
+    NSMutableDictionary<NSString*, QLPreviewReplyAttachment*>* attachments = [NSMutableDictionary dictionary];
+
+    NSString* name = @(metainfo.name().c_str());
+
+    auto const n_files = metainfo.file_count();
+    auto const is_multifile = n_files > 1;
+    UTType* fileType = is_multifile ? UTTypeFolder : [UTType typeWithFilenameExtension:name.pathExtension];
+
+    NSUInteger const width = 32;
+    [htmlString appendFormat:@"<h2><img class=\"icon\" src=\"%@\" width=\"%ld\" height=\"%ld\" />%@</h2>",
+                             generateIconData(fileType, width, attachments),
+                             width,
+                             width,
+                             name];
+
+    NSString* fileSizeString = [NSString stringForFileSize:metainfo.total_size()];
+    if (is_multifile)
+    {
+        NSString* fileCountString = [NSString
+            localizedStringWithFormat:NSLocalizedStringFromTableInBundle(@"%lu files", nil, bundle, "quicklook file count"), n_files];
```

Comment:
I believe this should be in a stringsdict file to correctly localise numeric amounts? (e.g. "1 file", "2 files"). Like described here: https://developer.apple.com/documentation/xcode/localizing-strings-that-contain-plurals

Reply:
I understand, what you mean here, but I'm very hesitant to update things here.

Most of this code is a copy of legacy QuickLook plugin, and all localizations are copied from there as is also.

So I don't want to change this in this PR, but I'm okay with updating this in consequent PRs, when new QuickLook Preview extension is shipped and we can introduce new strings to localize.